### PR TITLE
Adjust PrintMembers to avoid boxing and avoid extra space

### PR DESCRIPTION
**Pull request by jcouv**
Fixes https://github.com/dotnet/roslyn/issues/47093
Fixes https://github.com/dotnet/roslyn/issues/47092

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordPrintMembers.cs [hunk 1]
```diff
@@ -171,10 +171,21 @@ internal override void GenerateMethodBody(TypeCompilationState compilationState,
                         _ => throw ExceptionUtilities.UnexpectedValue(member.Kind)
                     };
 
-                    block.Add(F.ExpressionStatement(
-                        F.Call(receiver: builder,
-                            F.WellKnownMethod(WellKnownMember.System_Text_StringBuilder__AppendObject),
-                            F.Convert(F.SpecialType(SpecialType.System_Object), value))));
+                    Debug.Assert(value.Type is not null);
+                    if (value.Type.IsValueType)
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordPrintMembers.cs:175]: does this work for thigns like `where T : struct` type parameters?

Comment (jcouv) [reply src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordPrintMembers.cs:175]: it does. I'll add an explicit test

https://github.com/dotnet/roslyn/blob/master/src/Compilers/CSharp/Portable/Symbols/TypeParameterSymbol.cs#L564-L575

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordPrintMembers.cs [hunk 2]
```diff
@@ -157,7 +157,7 @@ internal override void GenerateMethodBody(TypeCompilationState compilationState,
                 {
                     // builder.Append(<name>);
                     // builder.Append(" = ");
-                    // builder.Append((object)<value>);
+                    // builder.Append(<value>.ToString()); OR builder.Append(<value>.ToString());
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordPrintMembers.cs:160]: ><value>.ToString() [](start = 77, length = 18)

`(object)<value>`

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordToString.cs [hunk 3]
```diff
@@ -65,11 +65,11 @@ internal override void GenerateMethodBody(TypeCompilationState compilationState,
                 // builder.Append(" { ");
                 block.Add(makeAppendString(F, builderLocal, " { "));
 
-                // this.print(builder);
-                block.Add(F.ExpressionStatement(F.Call(F.This(), _printMethod, builderLocal)));
+                // if (this.print(builder)) builder.Append(" ");
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordToString.cs:68]: >// if (this.print(builder)) builder.Append(" "); [](start = 16, length = 48)

If `PrintMembers()` is only used internally, to implement `ToString()`, could this `if (this.print(builder))` be removed and simply have `PrintMembers()` add a " " before each member?

Comment (cston) [reply src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordToString.cs:68]: > could this `if (this.print(builder))` be removed and simply have `PrintMembers()` add a " " before each member?

Given that `PrintMembers()` or `ToString()` can be implemented explicitly, it seems either requiring a trailing space here, or requiring a leading space before each member, will make one of the two methods somewhat awkward.

---
In reply to: [476183479](https://github.com/dotnet/roslyn/pull/47095#discussion_r476183479) [](ancestors = 476183479)

Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordToString.cs:68]: >print [](start = 28, length = 5)

Consider using `this.PrintMembers` explicitly in the comment.

**Comment by cston** [comment]
>     /// The `bool PrintMembers(StringBuilder)` method is responsible for printing members declared

Should `PrintMembers()` be called `AppendMembers()` (since it appends to a builder rather than printing)?

---
Refers to: src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordPrintMembers.cs:16 in 9fb3f77. [](commit_id = 9fb3f770847d75230f03c99d42ada45699b91019, deletion_comment = False)

**Comment by jcouv** [comment]
> Should PrintMembers() be called AppendMembers() (since it appends to a builder rather than printing)?

I don't think we can change the name now, as RC1 locks the public contract for records.

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordPrintMembers.cs [hunk 4]
```diff
@@ -157,7 +157,7 @@ internal override void GenerateMethodBody(TypeCompilationState compilationState,
                 {
                     // builder.Append(<name>);
                     // builder.Append(" = ");
-                    // builder.Append((object)<value>);
+                    // builder.Append((object)<value>); OR builder.Append(<value>.ToString());
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordPrintMembers.cs:160]: Consider clarifying when the `OR` applies.

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordToString.cs (hunk 3 above)
Comment (333fred) [reply src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordToString.cs]: I feel like the simpler implementation, for both us and consumers, is to just require a leading space in front of all members. Then there's no conditionals or questions.

---
In reply to: [476188140](https://github.com/dotnet/roslyn/pull/47095#discussion_r476188140) [](ancestors = 476188140,476183479)

**Comment by 333fred** [comment]
Done review pass (commit 5)

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordToString.cs (hunk 3 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordToString.cs]: I think it's more regular to let the caller of `PrintMembers` to deal with the returned bool and using it to print the separators. We do that for printing the comma separator; the space separator should be no different.

---
In reply to: [476650024](https://github.com/dotnet/roslyn/pull/47095#discussion_r476650024) [](ancestors = 476650024,476188140,476183479)

**Comment by RikkiGibson** [comment]
FYI, the correctness failure has been fixed in the base branch.

**Pull request merged**